### ISSUE TITLE
restapi should be in terraform registery as provider

### ISSUE THREAD

**Issue by thatsk** [Mastercard/terraform-provider-restapi#85]
https://www.terraform.io/docs/registry/modules/publish.html

**Comment by DRuggeri** [comment]
Thanks - good call! Looks like access to the registry is currently in closed beta, but we could definitely move the docs into the expected format documented here:
https://www.terraform.io/docs/registry/providers/docs.html

... it definitely would be nice to make the README a bit more 'sparse' and instead refer to the resource/datasource docs.

I'll tag this as help-wanted in case anyone has the interest. Otherwise I'll tackle it during a rainy day after the registry is open beyond private beta.

**Comment by paultyng** [comment]
👋 Hi, I'm on the Terraform Providers / Registry team at HashiCorp. Happy to help you get in to the beta if interested. We are also working on some tooling to help get documentation in to the registry format, but probably won't be ready for a little bit still.

To get invited to the closed beta, please email terraform-registry-beta@hashicorp.com. We need:

- A list of GitHub usernames to add to the beta (you and any team members who will publish)
- List of provider repositories you plan to publish
- A GPG public key in ASCII-armor format, which you will be using to sign your provider releases

You can use one key for all of your providers, or separate keys if you prefer. If you are publishing from an organization, this key or keys will be associated with that namespace. Once in the beta, you can manage personal keys in the UI as well.

**Comment by rquadling** [comment]
Hi.

Really would like to use this provider, but I'm not a Go dev and so some aspects of that side of things I'm not best suited to do any work on.

I've published 1 module into the registry, but never a provider.

@paultyng in terms of things like documentation, is there a tool that can extract the required content from the codebase? I use https://github.com/antonbabenko/pre-commit-terraform for the module and this is working well in our setup.

If you can give some pointers, then I can at least submit some PRs to the project and hopefully get this project more suitable for residing in the registry.

**Comment by paultyng** [comment]
We haven't formally announced it yet, but here is a tool that we are working on to generate documentation: https://github.com/hashicorp/terraform-plugin-docs

It requires you to define `Description` values in your schema information.

**Comment by paultyng** [comment]
Also just FYI, publishing on the registry is now open beta, but organization repos still need to send us their GPG keys manually (via terraform-registry@hashicorp.com)

**Comment by rquadling** [comment]
Ah. First hurdle ...

```
rendering website for provider "terraform-provider-restapi"
exporting schema from Terraform
compiling provider "restapi"
Error executing command: unable to generate website: 
Error: Failed to install provider

Error while installing hashicorp/restapi v0.0.1: provider
registry.terraform.io/hashicorp/restapi 0.0.1 is not available for
darwin_amd64
```

Sort of a chicken-and-egg situation!

If this issue is the sort of thing a Go dev would know how to solve straight away, then I'll leave it to them. But if this is one of those "Oh ... never considered that ... let me fix it" things ... I'm here to help.

**Comment by paultyng** [comment]
😂 ha, maybe I should test this locally first, I'll give it a shot.

**Comment by rquadling** [comment]
OOI, I tried running `go get` ...

```
~/go $ go get github.com/Mastercard/terraform-provider-restapi
package github.com/hashicorp/hcl/v2/hclwrite: cannot find package "github.com/hashicorp/hcl/v2/hclwrite" in any of:
	/usr/local/Cellar/go/1.14.5/libexec/src/github.com/hashicorp/hcl/v2/hclwrite (from $GOROOT)
	/Users/rquadling/go/src/github.com/hashicorp/hcl/v2/hclwrite (from $GOPATH)
package github.com/hashicorp/hcl/v2/json: cannot find package "github.com/hashicorp/hcl/v2/json" in any of:
	/usr/local/Cellar/go/1.14.5/libexec/src/github.com/hashicorp/hcl/v2/json (from $GOROOT)
	/Users/rquadling/go/src/github.com/hashicorp/hcl/v2/json (from $GOPATH)
package github.com/hashicorp/terraform/version: cannot find package "github.com/hashicorp/terraform/version" in any of:
	/usr/local/Cellar/go/1.14.5/libexec/src/github.com/hashicorp/terraform/version (from $GOROOT)
	/Users/rquadling/go/src/github.com/hashicorp/terraform/version (from $GOPATH)
```

So maybe there's some declaration/dependency not setup in an appropriate way? (guesswork from me!)

**Comment by rquadling** [comment]
Ah. I've now got the `~/go/bin/terraform-provider-restapi` file. Seems I had a (ahem) broken `~/go/src` directory (funny that!).

Hmm. Just out of my depth now as the building of the bin is there, but the plugin docs not working so much.

**Comment by MaxAnderson95** [comment]
Any update on this? Having this provider in the registry would be awesome!

**Comment by MaxAnderson95** [comment]
Based on the comments in #94 , it appears this issue may have been closed accidentally?

**Comment by DRuggeri** [comment]
Correct - I definitely didn't mean to close this out. *facepalm*
I guess that happened when I closed the doco PR automagically.

**Comment by DRuggeri** [comment]
So, @paultyng - I was looking to upload this module to the registry since releasing `v0.14.0` yesterday, but I'm a little unsure of the R/W access being requested for webhooks. Is there a way to do this without granting repo webhook/service r/w access to Hashicorp via my GitHub ID?

**Comment by paultyng** [comment]
Currently no, it writes the webhook with a shared secret that is generated just for your provider. We do have longer term plans to maybe allow for manual webhook management, but its still a ways down the backlog.

Not sure if you already had the registry app authorized for module usage from older than a few months ago, but if you had an old authorization, we have limited it a bit more than it used to have requested.

**Comment by chuckyz** [comment]
Looks like @fmontezuma forked this and set it up in their repo: https://github.com/fmontezuma/terraform-provider-restapi

**Comment by Unichron** [comment]
Any update on this?

**Comment by tbugfinder** [comment]
Hi @DRuggeri ,
would you find a minute and update this issue with the way forward? Should we rather use the helper repo of fmontezuma - short term?

**Comment by DRuggeri** [comment]
Sure - sorry, I guess I misunderstood what was shared in the note from @chuckyz . I had assumed that now that @fmontezuma has pushed this into the Terraform registry and claimed the "restapi" name, I am not able to do so. I could be wrong... so we'll see as I make progress.

I've gone ahead and started the process - granting access to appropriate portions as noted in prior comments, etc, but have yet to complete the process of establishing a GPG key that can be shared across those that develop/work on the provider and may push releases.

More to come...

**Comment by paultyng** [comment]
Just some context, provider names in 0.13 now have "namespaces", so `fmontezuma/restapi` is distinct from `Mastercard/restapi` in the registry (and in state, you could use them side by side). All historical ones were `hashicorp/*` or `terraform-providers/*` (sometimes this is shown in state as `-/*`).

**Comment by yi2020** [comment]
You can still claim yours as far as I know. I'm using:
`terraform {
  required_providers {
    restapi = {
      source  = "fmontezuma/restapi"
      version = "~> 1.14.0"
    }
  }
}
`

But am sure we can replace fmontezuma with mastercard above.

**Comment by colltoaction** [comment]
Any chance you could integrate https://github.com/Mastercard/terraform-provider-restapi/pull/114?

**Comment by gschaffer-cxn** [comment]
@DRuggeri any updates you can share?

**Comment by sticky-note** [comment]
@DRuggeri  Please, finish this

**Comment by blacksd** [comment]
@DRuggeri has there been any progress on this? Please, complete the work so we can all fetch the data from Terraform's registry. Thank you.

**Comment by arruko** [comment]
Heya @DRuggeri :) are there any limitations with the publication on the registry or do you have an ETA on when it will be released there? This is to evaluate whether to continue using the provider or not at least until it's there

**Comment by crosbymichael1** [comment]
To all waiting for the official provider to be added:

Ive been using this setup for 5 months now and its been working perfectly, so if anyone is in desperate need of it I'd recommend this:

```terraform
terraform {
  required_providers {
    restapi = {
      source = "fmontezuma/restapi"
      version = "1.14.1"
    }
  }
}
```

**Comment by TryTryAgain** [comment]
Same

**Comment by DRuggeri** [comment]
Thank you, all, for your patience. @blz-ea opened up #114 with a PR that gets this so close! We'll definitely go the GitHub actions route - but I could use a hand getting this across the line. There are a few notes in the [PR](https://github.com/Mastercard/terraform-provider-restapi/pull/114#issuecomment-805004359) - if anyone has a moment to help rejigger the proposed pipeline to use raw commands rather than an opaque action, I think we can get this thing done!

**Comment by blz-ea** [comment]
Hey guys I've opened a new PR (#127) as continuation of #114 with (probably) all notes addressed.

**Comment by DRuggeri** [comment]
OK, all - I could use a few testers/verification after merging in #127
Can folks please take a look over here in my sandbox?
https://github.com/DRuggeri/terraform-provider-restapi/releases/tag/v1.16.0

One thing I'm wondering about is whether this release may offend the registry in some way because it includes fakeserver files. Would you happen to know if that matters to the registry, @paultyng?

**Comment by paultyng** [comment]
I no longer work for HashiCorp, so not familiar with any new processes, the only files the registry was looking at previously were the assets on the release and documents on the commit for the release.

**Comment by wb529kq** [comment]
@DRuggeri 
Partner and community providers are signed by their developers.
If you'd like to know more about provider signing, you can read about it here:

https://www.terraform.io/docs/cli/plugins/signing.html
Error: Failed to install provider

Error while installing blz-ea/restapi v0.0.3: could not query provider
registry for registry.terraform.io/blz-ea/restapi: failed to retrieve
authentication checksums for provider: 404 Not Found

**Comment by emmaroberts-nbs** [comment]
Any update on this please?

**Comment by emmaLP** [comment]
I am happy to help get this sorted it.

**Comment by scastria** [comment]
yes, please can someone address this issue.  Many people are forking this repo and then publishing it as their own so there are maybe 5 copies of this provider in the registry each forked at different points in history.  It is very confusing.

**Comment by ViacheslavKudinov** [comment]
There is a step by step guide how to release provider
https://learn.hashicorp.com/tutorials/terraform/provider-release-publish?in=terraform/providers
It will be great finally get official rest api provider.

**Comment by FalcoSuessgott** [comment]
what is the problem here? why is it still not published?

**Comment by DRuggeri** [comment]
I've asked for testing in [this comment](https://github.com/Mastercard/terraform-provider-restapi/issues/85#issuecomment-821827406) to see if the pipeline tweaks I made work... but have yet to hear anyone from the community having given things a whirl. It's important to me that someone can independently verify that things work as expected. Can *anyone* do this?

**Comment by wb529kq** [comment]
# I can happily help test but do not know how you want us to test I cannot find it in the registry

**Comment by DRuggeri** [comment]
Well, I messed with it again for a while and since there are just too many nice things to have (several community contributions, fixes, and misc stuff) I went ahead and just updated the GitHub release workflow and `goreleaser` configs, then enrolled the result in the registry.

For now the registry simply says:
```
 Could not find all platform assets for this release yet (1.16.0) — 3 minutes ago

missing release asset for ["fakeserver_1.16.0_linux_amd64.zip" "fakeserver_1.16.0_linux_386.zip" "fakeserver_1.16.0_windows_amd64.zip" "fakeserver_1.16.0_windows_arm.zip" "fakeserver_1.16.0_freebsd_386.zip" "fakeserver_1.16.0_freebsd_arm64.zip" "fakeserver_1.16.0_linux_arm.zip" "fakeserver_1.16.0_darwin_amd64.zip" "fakeserver_1.16.0_freebsd_amd64.zip" "fakeserver_1.16.0_freebsd_arm.zip" "fakeserver_1.16.0_windows_386.zip" "fakeserver_1.16.0_linux_arm64.zip"] 
```
... but I see that all such assets are present as well as the sigs. So, maybe it just needs a while to catch up and integrate everything.

**Comment by ViacheslavKudinov** [comment]
@DRuggeri maybe this could help

> If your version doesn't appear properly, you may force a sync with GitHub by viewing your module on the registry and clicking "Resync Module" under the "Manage Module" dropdown. This process may take a few minutes. Please only do this if you do not see the version appear, since it will cause the registry to resync all versions of your module.

Was mentioned there https://www.terraform.io/docs/registry/modules/publish.html#releasing-new-versions

**Comment by serialseb** [comment]
@DRuggeri just an idea looking for 5 minutes, do you think maybe the signature files should exist for both the provider and for fakeserver, as two different set of file prefixes may result in two sets of signature files being looked at?

**Comment by ViacheslavKudinov** [comment]
It will be soon 2 years as the issue with publishing is there.
@DRuggeri can you grant someone permissions to help you with this process?
Maybe, some people had experience with the process and can easily achieve it. At least, I hope.

**Comment by DRuggeri** [comment]
Hi, @ViacheslavKudinov - I don't think the problem is a permission issue. I think, based on the error message, nobody seems to know what to do. If I could get some guidance from Hashicorp, I'm certain that I can resolve whatever the issue is. The ticket I opened has still yet to be acknowledged despite followup :-(

I'm wondering if I just need to bundle fakeserver as though it's a release artifact of the provider itself (which... it definitely isn't) given the [error message](https://github.com/Mastercard/terraform-provider-restapi/issues/85#issuecomment-913875579) - 
 but I suspect that would lead to other failures since fakeserver is clearly not restapi.

**Comment by scastria** [comment]
I have 2 providers in the registry.  I followed the directions [here](https://www.terraform.io/registry/providers/publishing)

I see that your publishing related files have modifications that include fakeserver and stuff.  Do you guys use the .goreleaser.yml locally OR is it only there to try to publish this to the registry?  If you use it locally, then maybe you need the fakeserver stuff in there, but that is causing issues with your registry publishing.  I could submit a PR that changes the publishing related files back to the standard from this [repo](https://github.com/hashicorp/terraform-provider-scaffolding)

**Comment by scastria** [comment]
The reason I am now motivated to help is that I just got a Mac with the Apple M1 chip which requires a darwin_arm64 binary from this provider.  I have been using kirillmeleshko/restapi fork of this provider but he hasn't updated his fork to add darwin_arm64 support, which is a one character change.  But I am tired of using forks, this should be published from here.

**Comment by scastria** [comment]
Screw it.  I guess I will add to the 143 forks with my own fork so that I can add darwin_arm64 support.

**Comment by DRuggeri** [comment]
Hi, @scastria - I understand your point, but it is not as straight forward as you envision. The problem, as noted a few times in the messages posted on the topic, is that the Hashicorp registry is rejecting the release - not because it's incorrect, but because it is looking for files that it should not look for. The GitHub actions and goreleaser configs came straight from the latest/greatest Hashicorp recommendation. The only difference of substance is that we (this project) is also building the fakeserver binary with goreleaser. That, I believe, is where/how it's having issues.

A possible solution would be to just stop building/distributing fakeserver. I'm not opposed to trying that, so am happy to entertain a PR and updated tests that accomplish this if anyone has the time. Otherwise, I will get to it to try things out and see how it goes.

**Comment by scastria** [comment]
I don't know anything about GitHub actions so would not be able to figure out what is wrong with your current goreleaser or what is wrong with how TF expects for releases to be distributed.  However, I do know that if I:

1. Replace release.yml in .github/workflows with the latest copy from scaffolding example repo
2. Replace .goreleaser.yml with latest copy from scaffolding example repo
3. Add a terraform-registry-manifest.json file with latest copy from scaffolding example repo
4. Increase Go version in release.yml to add support for darwin_arm64 platform
5. Publish

that every works great!

My new restapi published provider with darwin_arm64 support is [here](https://registry.terraform.io/providers/scastria/restapi/latest)

I am sure that fakeserver is very important for unit testing.  However, everybody that wants to consume this TF provider straight from the author without having to go through some random guy's fork (like mine) doesn't care about fakeserver.

**Comment by scastria** [comment]
@DRuggeri I see that you deleted one of my comments above.  I can understand that as I wrote that out of frustration.  But, I hope you can understand everybody's frustration here.  I was able to release your provider in 1hr 27min from start to finish, and yet, this issue has been open for 625 days.

**Comment by DRuggeri** [comment]
No problem - I understand the frustration. Thank you for taking the time to open a PR with what worked for you - this is incredibly helpful.

I have to agree that fakeserver just needs to be abandoned from the release. Without any further input from Hashicorp, this change at least unblocks publishing the to registry.

**Comment by DRuggeri** [comment]
According to [the registry](https://registry.terraform.io/providers/Mastercard/restapi/1.16.1), the module has been published!

I assume we're good to (finally) close the issue. Thank you, all, that provided guidance and help!
